hook-janitor/src/webhooks.rs

```diff
@@ -1,55 +1,454 @@
+    // ever fails then something has gone wrong.
+    #[sqlx(try_from = "i64")]
+    team_id: u32,
+    #[sqlx(try_from = "i64")]
+    plugin_config_id: u32,
+    #[sqlx(json)]
+    last_error: WebhookJobError,
+    #[sqlx(try_from = "i64")]
+    failures: u32,
 }
 
 impl WebhookCleaner {
     pub fn new(
         queue_name: &str,
         table_name: &str,
         database_url: &str,
-        batch_size: u32,
         kafka_producer: FutureProducer<KafkaContext>,
         app_metrics_topic: String,
-        plugin_log_entries_topic: String,
-    ) -> Result<Self, CleanerError> {
+    ) -> Result<Self> {
         let queue_name = queue_name.to_owned();
         let table_name = table_name.to_owned();
         let pg_pool = PgPoolOptions::new()
+            .acquire_timeout(Duration::from_secs(10))
             .connect_lazy(database_url)
-            .map_err(|error| CleanerError::PoolCreationError { error })?;
+            .map_err(|error| WebhookCleanerError::PoolCreationError { error })?;
 
         Ok(Self {
             queue_name,
             table_name,
             pg_pool,
-            batch_size,
             kafka_producer,
             app_metrics_topic,
-            plugin_log_entries_topic,
         })
     }
+
+    #[allow(dead_code)] // This is used in tests.
+    pub fn new_from_pool(
+        queue_name: &str,
+        table_name: &str,
+        pg_pool: PgPool,
+        kafka_producer: FutureProducer<KafkaContext>,
+        app_metrics_topic: String,
+    ) -> Result<Self> {
+        let queue_name = queue_name.to_owned();
+        let table_name = table_name.to_owned();
+
+        Ok(Self {
+            queue_name,
+            table_name,
+            pg_pool,
+            kafka_producer,
+            app_metrics_topic,
+        })
+    }
+
+    async fn start_serializable_txn(&self) -> Result<Transaction<'_, Postgres>> {
+        let mut tx = self
+            .pg_pool
+            .begin()
+            .await
+            .map_err(|e| WebhookCleanerError::StartTxnError { error: e })?;
+
+        // We use serializable isolation so that we observe a snapshot of the DB at the time we
+        // start the cleanup process. This prevents us from accidentally deleting rows that are
+        // added (or become 'completed' or 'failed') after we start the cleanup process.
+        //
+        // If we find that this has a significant performance impact, we could instead move
+        // rows to a temporary table for processing and then deletion.
+        sqlx::query("SET TRANSACTION ISOLATION LEVEL SERIALIZABLE")
+            .execute(&mut *tx)
+            .await
+            .map_err(|e| WebhookCleanerError::StartTxnError { error: e })?;
+
+        Ok(tx)
+    }
+
+    async fn get_completed_rows(
+        &self,
+        tx: &mut Transaction<'_, Postgres>,
+    ) -> Result<Vec<CompletedRow>> {
+        let base_query = format!(
+            r#"
+            SELECT DATE_TRUNC('hour', finished_at) AS hour,
```

Comment:
Doesn't this mean that if we ever change the aggregation on app metrics' side, we will be losing information here? For example, if we ever start trunc-ing by 15 minutes on app metrics, we will have already sent truncated dates that provide no precision for 15 minutes aggregates.

Reply:
Ugh, but I can't think of a better alternative, so maybe we just live with this.

Reply:
That is true, but it seems like a big thing to change the aggregation level of the `app_metrics` table. I would think we'd make a new topic and revisit producers to it, but I guess that's a big assumption I'm making.

I don't have to do any aggregation here, but it seems like a shame to not, since the _only_ reason we're reading the rows is to write to `app_metrics`.

How about I PR to PostHog to document that the aggregation is import to existing producers like rusty-hook?

Reply:
https://github.com/PostHog/posthog/pull/19449